### PR TITLE
Treat untyped blank as a scalar rather than a table

### PR DESCRIPTION
**Pull request by gregli-msft**
Addresses https://github.com/microsoft/Power-Fx/issues/1317 based on excellent suggestions from Carlos.  We will now match Canvas semantics and lean toward an untyped blank being a scalar rather than a table.

### CODE REVIEW

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/UntypedBlankAsTable.txt [hunk 1]
```diff
@@ -1,42 +1,47 @@
...
 Table({Value:"bc"},{Value:"ef"})
 
 >> Right(["abc","def"],Blank())
-#SKIP: Waiting on https://github.com/microsoft/Power-Fx/issues/1317, JS and C# disagree
+Table({Value:""},{Value:""})
 
 >> Right("abc",[2,3])
 Table({Value:"bc"},{Value:"abc"})
 
+// suspect, may be different on Canvas requiring an override
```
Comment (gregli-msft) [inline src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/UntypedBlankAsTable.txt:42]: There are three tests that give different results in Canvas.  We need to validate that these are the right answers and that Canvas should snap to these semantics with Power Fx 1.0.   @CarlosFigueiraMSFT

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/UntypedBlankAsTable.txt [hunk 2]
```diff
@@ -1,42 +1,47 @@
...
 Table({Value:"ab"},{Value:"de"})
 
 >> Left(["abc","def"],Blank())
-#SKIP: Waiting on https://github.com/microsoft/Power-Fx/issues/1317, JS and C# disagree
+Table({Value:""},{Value:""})
 
 >> Left("abcd",[2,3])
 Table({Value:"ab"},{Value:"abc"})
 
+// suspect, may be different on Canvas requiring an override
```
Comment (gregli-msft) [inline src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/UntypedBlankAsTable.txt:29]: There are three tests that give different results in Canvas.  We need to validate that these are the right answers and that Canvas should snap to these semantics with Power Fx 1.0.  @CarlosFigueiraMSFT

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/TabularOverloadsBlanksAndErrors.txt [hunk 3]
```diff
@@ -252,10 +251,11 @@ Table({Value:0},{Value:0},{Value:0})
 Table()
 
 >> Concatenate(Blank(), Filter([1], Value > 10))
-Blank()
+Table()
 
 >> Concatenate(Blank(), "test")
 "test"
 
+// suspect, may be different on Canvas requiring an override
```
Comment (gregli-msft) [inline src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/TabularOverloadsBlanksAndErrors.txt:259]: There are three tests that give different results in Canvas.  We need to validate that these are the right answers and that Canvas should snap to these semantics with Power Fx 1.0.  @CarlosFigueiraMSFT

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/TabularOverloadsBlanksAndErrors.txt (hunk 3 above)
Comment (CarlosFigueiraMSFT) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/TabularOverloadsBlanksAndErrors.txt:259]: Checking why this doesn't work properly in canvas.

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/UntypedBlankAsTable.txt (hunk 1 above)
Comment (CarlosFigueiraMSFT) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/UntypedBlankAsTable.txt:42]: This works as expected in Power Apps

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/UntypedBlankAsTable.txt (hunk 2 above)
Comment (CarlosFigueiraMSFT) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/UntypedBlankAsTable.txt:29]: This works as expected in Power Apps

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/FindT.txt [hunk 4]
```diff
@@ -7,37 +6,37 @@ Errors: Error 0-24: The function 'Find' has some invalid arguments.|Error 14-18:
 Errors: Error 0-25: The function 'Find' has some invalid arguments.|Error 5-9: Invalid argument type (Table). Expecting a Text value instead.|Error 11-15: Invalid argument type (Table). Expecting a Text value instead.
 
 >> Find(If(false, ["blank table"], Blank()), If(false, ["blank table"], Blank()), [1])
-Blank()
+Table({Value:1})
```
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/FindT.txt:9]: This is not right... The first (and second) argument is a blank table. For tabular functions, if a blank table is input, then the result should be blank

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/UntypedBlankAsTable.txt (hunk 1 above)
Comment (gregli-msft) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/UntypedBlankAsTable.txt]: My fault, I neglected to turn on Excel Compat.  The test is OK.

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/UntypedBlankAsTable.txt (hunk 2 above)
Comment (gregli-msft) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/UntypedBlankAsTable.txt]: My fault, I neglected to turn on Excel Compat.  The test is OK.

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/TabularOverloadsBlanksAndErrors.txt (hunk 3 above)
Comment (gregli-msft) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/TabularOverloadsBlanksAndErrors.txt]: C# has the correct behavior.  Carlos is looking into why Canvas is different.

**Comment by LucGenetier** [comment]
✅ No public API change.

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/FindT.txt (hunk 4 above)
Comment (gregli-msft) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/FindT.txt]: Resolved by recent Carlos changes.

**Comment by LucGenetier** [comment]
✅ No public API change.

**Comment by Shpakh-MSFT** [comment]
How would this work for below expression? Would return type for below expression be Void?
If(false, [1,2], Blank())

**Comment by CarlosFigueiraMSFT** [comment]
This change affects mostly tabular functions - the logic to determine the return type of If is not changed. In this case, since *[Value:n] _accepts_ ObjNull (the type of Blank()), the result will be *[Value:n].

---
In reply to: [1624171294](https://github.com/microsoft/Power-Fx/pull/1635#issuecomment-1624171294) [](http://example.com/codeflow?ancestors=1624171294,1624171294)

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/libraries/Microsoft.PowerFx.Core/Types/DType.cs [hunk 5]
```diff
@@ -578,6 +578,8 @@ internal void AssertValid()
 
         public bool IsTable => Kind == DKind.Table || Kind == DKind.ObjNull || Kind == DKind.LazyTable;
 
+        public bool IsTableStrict => Kind == DKind.Table || Kind == DKind.LazyTable;
```
Comment (bryancroteau-MSFT) [inline src/libraries/Microsoft.PowerFx.Core/Types/DType.cs:581]: In the future, we _might_ have to add a Boolean flag with Kind == DKind.Table || Kind == DKind.ObjNull without the LazyTable one. Then, isn't that one also "Strict?" We are currently not using the word "Strict" anywhere in this document. Could we make it more descriptive? Maybe IsNonBlankTable?

Comment (CarlosFigueiraMSFT) [reply src/libraries/Microsoft.PowerFx.Core/Types/DType.cs]: Updated it to `IsTableNonObjNull`

**Review by bryancroteau-MSFT — APPROVED** [review]

**Review by Shpakh-MSFT — CHANGES_REQUESTED** [review]
:clock1:

**Comment by LucGenetier** [comment]
✅ No public API change.

**Review by Shpakh-MSFT — APPROVED** [review]
:shipit:

**Pull request merged**